integrations/lib/multiwoven/integrations/source/aws_athena/client.rb

```diff
@@ -0,0 +1,121 @@
+        query = "SELECT table_name, column_name, data_type, is_nullable FROM information_schema.columns WHERE table_schema = '#{connection_config[:schema]}' ORDER BY table_name, ordinal_position;"
+        db = create_connection(connection_config)
+        response = db.start_query_execution(
+          query_string: query,
+          query_execution_context: { database: connection_config[:schema] },
+          result_configuration: { output_location: connection_config[:output_location] }
+        )
+        query_execution_id = response[:query_execution_id]
+        wait_for_query_completion(db, query_execution_id)
+
+        results = transform_results(db.get_query_results(query_execution_id: query_execution_id))
+        catalog = Catalog.new(streams: create_streams(results))
+        catalog.to_multiwoven_message
+      rescue StandardError => e
+        handle_exception(
+          "AWS:ATHENA:DISCOVER:EXCEPTION",
+          "error",
+          e
+        )
+      end
+
+      def read(sync_config)
+        connection_config = sync_config.source.connection_specification
+        connection_config = connection_config.with_indifferent_access
+        query = sync_config.model.query
+        query = batched_query(query, sync_config.limit, sync_config.offset) unless sync_config.limit.nil? && sync_config.offset.nil?
+
+        db = create_connection(connection_config)
+        response = db.start_query_execution(
+          query_string: query,
+          query_execution_context: { database: sync_config[:source][:connection_specification][:schema] },
+          result_configuration: { output_location: sync_config[:source][:connection_specification][:output_location] }
+        )
+        query_execution_id = response[:query_execution_id]
+        wait_for_query_completion(db, query_execution_id)
+        results = transform_results(db.get_query_results(query_execution_id: query_execution_id))
+        query(results)
+      rescue StandardError => e
+        handle_exception(
+          "AWS:ATHENA:READ:EXCEPTION",
+          "error",
+          e
+        )
+      end
+
+      private
+
+      def create_connection(connection_config)
+        Aws.config.update({ credentials: Aws::Credentials.new(connection_config[:access_key], connection_config[:secret_access_key]), region: connection_config[:region] })
+        Aws::Athena::Client.new
+      end
+
+      def create_streams(records)
+        group_by_table(records).map do |_, r|
+          Multiwoven::Integrations::Protocol::Stream.new(name: r[:tablename], action: StreamAction["fetch"], json_schema: convert_to_json_schema(r[:columns]))
+        end
+      end
+
+      def wait_for_query_completion(db, query_execution_id)
+        loop do
+          response = db.get_query_execution(query_execution_id: query_execution_id)
+          status = response.query_execution.status.state
+          break if %w[SUCCEEDED FAILED CANCELLED].include?(status)
+
+          sleep 1
```

Comment:
We refrain from using sleep.  Why do we need sleep and also do we need a loop?

Reply:
As for the sleep sometimes the get_query_execution keeps running instead of awaiting for a response. Will look into how to fix the problem without sleep